cpp/src/copying/contiguous_split.cu

```diff
@@ -193,11 +193,12 @@ __device__ void copy_buffer(uint8_t* __restrict__ dst,
     // and will never both be true at the same time.
     if (value_shift || bit_shift) {
       std::size_t idx = (num_bytes - remainder) / 4;
-      uint32_t v      = remainder > 0 ? (reinterpret_cast<uint32_t*>(src)[idx] - value_shift) : 0;
+      uint32_t v = remainder > 0 ? (reinterpret_cast<uint32_t const*>(src)[idx] - value_shift) : 0;
       while (remainder) {
-        uint32_t const next =
-          remainder > 0 ? (reinterpret_cast<uint32_t*>(src)[idx + 1] - value_shift) : 0;
-        uint32_t const val = (v >> bit_shift) | (next << (32 - bit_shift));
+        uint32_t const next = bit_shift > 0 || remainder > 4
```

Comment:
Can we not hoist out the conditional by stopping the loop early and doing cleanup?

```
while (remainder > 4) {
  uint32_t const next =
      remainder > 0 ? (reinterpret_cast<uint32_t *>(src)[idx + 1] - value_shift)
                    : 0;
  uint32_t const val = (v >> bit_shift) | (next << (32 - bit_shift));
  if (valid_count) {
    thread_valid_count += __popc(val);
  }
  reinterpret_cast<uint32_t *>(dst)[idx] = val;
  v = next;
  idx++;
  remainder -= 4;
}
if (remainder != 0) { // pad the remainder with 0's
  reinterpret_cast<uint32_t *>(dst)[idx] = 0;
}
```

Reply:
Unfortunately not. I tried this as well but it resulted in an incorrect null count. This code is doing double duty for copying/shifting offsets and bitmasks and each has different source boundaries. This was the only way to get both results to work. Well, without some total rewrite that specializes code for each.